virtcontainers/cgroups.go

```diff
@@ -0,0 +1,192 @@
+// Copyright (c) 2018 Huawei Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package virtcontainers
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/containerd/cgroups"
+	"github.com/kata-containers/runtime/virtcontainers/pkg/annotations"
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+)
+
+const (
+	vcpuGroupName       = "vcpu"
+	defaultCgroupParent = "/vc"
+)
+
+type sandboxCgroups struct {
+	commonParent cgroups.Cgroup
+	sandboxSub   cgroups.Cgroup
+	vcpuSub      cgroups.Cgroup
+}
+
+func (s *Sandbox) newCgroups() error {
+	// New will still succeed when cgroup exists
+	// create common parent for all kata-containers
+	// e.g. /sys/fs/cgroup/cpu/vc
+	parent, err := cgroups.New(cgroups.V1,
+		cgroups.StaticPath(defaultCgroupParent), &specs.LinuxResources{})
+	if err != nil {
+		return fmt.Errorf("failed to create cgroup for %q", defaultCgroupParent)
+	}
+
+	// create sub-cgroup for each sandbox
+	// e.g. /sys/fs/cgroup/cpu/vc/<sandbox>
+	sandboxSub, err := parent.New(s.id, &specs.LinuxResources{})
+	if err != nil {
+		return fmt.Errorf("failed to create cgroup for %s/%s", defaultCgroupParent, s.id)
+	}
+
+	// create sub-cgroup for vcpu threads
+	vcpuSub, err := sandboxSub.New(vcpuGroupName, &specs.LinuxResources{})
+	if err != nil {
+		return fmt.Errorf("failed to create cgroup for %s/%s/%s", defaultCgroupParent, s.id, vcpuGroupName)
+	}
+
+	s.cgroup = &sandboxCgroups{
+		commonParent: parent,
+		sandboxSub:   sandboxSub,
+		vcpuSub:      vcpuSub,
+	}
+	return nil
+}
+
+func (s *Sandbox) destroyCgroups() error {
+	// first move all processes in subgroup to parent in case live process blocks
+	// deletion of cgroup
+	if s.cgroup == nil {
+		// no need to destroy
+		s.Logger().Warningf("cgroup not initialized, no need to destroy")
+		return nil
+	}
+
+	if err := s.cgroup.sandboxSub.MoveTo(s.cgroup.commonParent); err != nil {
```

Comment:
@WeiZhang555 Can you add some comments explaining what you are doing here?

Reply:
ok